plugins/experimental/otel_tracer/otel_tracer.cc

```diff
@@ -0,0 +1,363 @@
+    }
+
+    TSHandleMLocRelease(buf, TS_NULL_MLOC, hdr_loc);
+  }
+
+  retval = 1;
+
+lReturn:
+  req_data->Destruct(req_data);
+  delete req_data;
+
+  TSContDestroy(contp);
+  TSHttpTxnReenable(txnp, TS_EVENT_HTTP_CONTINUE);
+
+  return retval;
+}
+
+static void
+set_request_header(TSMBuffer buf, TSMLoc hdr_loc, const char *key, int key_len, const char *val, int val_len)
+{
+  TSMLoc field_loc, tmp;
+  int first;
+
+  field_loc = TSMimeHdrFieldFind(buf, hdr_loc, key, key_len);
+
+  if (field_loc != TS_NULL_MLOC) {
+    first = 1;
+    while (field_loc != TS_NULL_MLOC) {
+      tmp = TSMimeHdrFieldNextDup(buf, hdr_loc, field_loc);
+      if (first) {
+        first = 0;
+        TSMimeHdrFieldValueStringSet(buf, hdr_loc, field_loc, -1, val, val_len);
+      } else {
+        TSMimeHdrFieldDestroy(buf, hdr_loc, field_loc);
+      }
+      TSHandleMLocRelease(buf, hdr_loc, field_loc);
+      field_loc = tmp;
+    }
+  } else if (TSMimeHdrFieldCreateNamed(buf, hdr_loc, key, key_len, &field_loc) != TS_SUCCESS) {
+    TSError("[otel_tracer][%s] TSMimeHdrFieldCreateNamed error", __FUNCTION__);
+    return;
+  } else {
+    TSMimeHdrFieldValueStringSet(buf, hdr_loc, field_loc, -1, val, val_len);
+    TSMimeHdrFieldAppend(buf, hdr_loc, field_loc);
+  }
+
+  if (field_loc != TS_NULL_MLOC) {
+    TSHandleMLocRelease(buf, hdr_loc, field_loc);
+  }
+
+  return;
+}
+
+static std::string
+read_request_header(TSMBuffer buf, TSMLoc hdr_loc, const char *key, int key_len)
+{
+  std::string retval = "";
+  const char *field;
+  int field_len;
+  TSMLoc field_loc;
+  field_loc = TSMimeHdrFieldFind(buf, hdr_loc, key, key_len);
+  if (field_loc) {
+    field = TSMimeHdrFieldValueStringGet(buf, hdr_loc, field_loc, -1, &field_len);
+    retval.assign(field, field_len);
+    TSHandleMLocRelease(buf, hdr_loc, field_loc);
+  }
+
+  return retval;
+}
+
+static void
+read_request(TSHttpTxn txnp, TSCont contp)
+{
+  TSMBuffer buf;
+  TSMLoc hdr_loc;
+
+  if (TSHttpTxnClientReqGet(txnp, &buf, &hdr_loc) != TS_SUCCESS) {
+    TSError("[otel_tracer][%s] cannot retrieve client request", __FUNCTION__);
+    TSHttpTxnReenable(txnp, TS_EVENT_HTTP_CONTINUE);
+    return;
+  }
+
+  // method
+  const char *method;
+  int method_len;
+  method                 = TSHttpHdrMethodGet(buf, hdr_loc, &method_len);
+  std::string method_str = "";
+  method_str.assign(method, method_len);
+
+  // TO-DO: add http flavor as attribute
+  // const char *h2_tag = TSHttpTxnClientProtocolStackContains(txnp, "h2");
+  // const char *h1_tag = TSHttpTxnClientProtocolStackContains(txnp, "http/1.0");
+  // const char *h11_tag = TSHttpTxnClientProtocolStackContains(txnp, "http/1.1");
+
+  // target
+  char *target           = nullptr;
+  int target_len         = 0;
+  std::string target_str = "";
+  target                 = TSHttpTxnEffectiveUrlStringGet(txnp, &target_len);
+  if (target) {
+    target_str.assign(target, target_len);
+    free(target);
```

Comment:
Shouldn't that be `TSFree`?
Further, it would be better to use `TextView` to map the returned value and free it on return from this method.